[bug] Seek past end of file

Calling `Reader.Seek` to a location past the end of the file results in a crash. If the calculated chunk # is greater than the number of actual chunks the following crash occurs.

```
panic: runtime error: index out of range [4] with length 2

goroutine 1 [running]:
github.com/ianlewis/go-dictzip.(*Reader).readChunk(0x5e0ac0?, 0xc0001876d8?, 0x46e8a9?)
        /home/ian/src/go-dictzip/reader.go:267 +0x30d
github.com/ianlewis/go-dictzip.(*Reader).Read(0xc0000a2c30, {0xc0001e2000, 0x8000, 0xc0000a0300?})
        /home/ian/src/go-dictzip/reader.go:226 +0x30
io.copyBuffer({0x668198, 0xc0000a0300}, {0x667fd8, 0xc0000a2c30}, {0x0, 0x0, 0x0})
        /home/ian/opt/go/src/io/io.go:429 +0x191
io.Copy(...)
        /home/ian/opt/go/src/io/io.go:388
os.genericReadFrom(0x5e0ac0?, {0x667fd8, 0xc0000a2c30})
        /home/ian/opt/go/src/os/file.go:185 +0x4f
os.(*File).ReadFrom(0xc0000a0090, {0x667fd8, 0xc0000a2c30})
        /home/ian/opt/go/src/os/file.go:161 +0x9c
io.copyBuffer({0x668038, 0xc0000a0090}, {0x667fd8, 0xc0000a2c30}, {0x0, 0x0, 0x0})
        /home/ian/opt/go/src/io/io.go:415 +0x151
io.Copy(...)
        /home/ian/opt/go/src/io/io.go:388
main.(*decompress).decompress(0xc000187ba0, {0x668038, 0xc0000a0090}, 0xc0000a02f8)
        /home/ian/src/go-dictzip/cmd/dictzip/decompress.go:127 +0x325
main.(*decompress).Run(0xc000187ba0)
        /home/ian/src/go-dictzip/cmd/dictzip/decompress.go:69 +0x438
main.newDictzipApp.func1(0xc0000b6340)
        /home/ian/src/go-dictzip/cmd/dictzip/app.go:231 +0x69f
github.com/urfave/cli/v2.(*Command).Run(0xc0001bc160, 0xc0000b6340, {0xc0000ba000, 0x5, 0x5})
        /home/ian/src/go-dictzip/vendor/github.com/urfave/cli/v2/command.go:276 +0x7e2
github.com/urfave/cli/v2.(*App).RunContext(0xc0000ca200, {0x6693e0, 0x838be0}, {0xc0000ba000, 0x5, 0x5})
        /home/ian/src/go-dictzip/vendor/github.com/urfave/cli/v2/app.go:333 +0x5a5
github.com/urfave/cli/v2.(*App).Run(...)
        /home/ian/src/go-dictzip/vendor/github.com/urfave/cli/v2/app.go:307
main.main()
        /home/ian/src/go-dictzip/cmd/dictzip/main.go:31 +0x3b
exit status 2
```

--- a/reader.go
+++ b/reader.go
@@ -264,6 +264,10 @@ func (z *Reader) Seek(offset int64, whence int) (int64, error) {
 // number of bytes advanced in the underlying reader and bytes read.
 func (z *Reader) readChunk(offset int64, size int) ([]byte, error) {
 	chunkNum := offset / int64(z.chunkSize)
+	if chunkNum >= int64(len(z.offsets)) {
+		// NOTE: We are trying to seek past the end of the file.
+		return nil, io.EOF
+	}
 	chunkOffset := z.offsets[chunkNum]
 
 	if _, err := z.r.Seek(chunkOffset, io.SeekStart); err != nil {

--- a/reader_test.go
+++ b/reader_test.go
@@ -515,6 +515,141 @@ func TestReader_Seek(t *testing.T) {
 	}
 }
 
+func TestReader_Seek_EOF_last_chunk(t *testing.T) {
+	t.Parallel()
+
+	// NOTE: uncompressed data: "chunk1chunk2chunk3last"
+	z, err := NewReader(bytes.NewReader([]byte{
+		// Header
+		hdrGzipID1,
+		hdrGzipID2,
+		hdrDeflateCM,
+		flgEXTRA,               // FLG
+		0x00, 0x00, 0x00, 0x00, // MTIME
+		0x0,       // XFL
+		OSUnknown, // OS
+
+		// EXTRA
+		0x12, 0x0, // XLEN // 18
+		0x52, 0x41, // 'R', 'A'
+		0xe, 0x0, // LEN // 14
+		0x1, 0x0, // VER // 1
+		0x6, 0x0, // CHLEN // 6
+		0x4, 0x0, // CHCNT // 4
+
+		// Chunk sizes.
+		0xc, 0x0, // 12
+		0xc, 0x0, // 12
+		0xc, 0x0, // 12
+		0x0a, 0x00, // 10
+
+		// compressed data (three 12 byte chunks and one 10 byte chunk).
+		0x4a, 0xce, 0x28, 0xcd, 0xcb, 0x36, 0x04, 0x00, 0x00, 0x00, 0xff, 0xff,
+		0x4a, 0xce, 0x28, 0xcd, 0xcb, 0x36, 0x02, 0x00, 0x00, 0x00, 0xff, 0xff,
+		0x4a, 0xce, 0x28, 0xcd, 0xcb, 0x36, 0x06, 0x00, 0x00, 0x00, 0xff, 0xff,
+		0xca, 0x49, 0x2c, 0x2e, 0x01, 0x00, 0x00, 0x00, 0xff, 0xff,
+
+		0x01, 0x00, 0x00, 0xff, 0xff, // sync/end marker.
+
+		0x70, 0xee, 0xf4, 0x09, // CRC-32
+		0x16, 0x00, 0x00, 0x00, // ISIZE // 22 (len of data)
+	}))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// NOTE: Seek past EOF but still within chunk size of last chunk.
+	off, err := z.Seek(22, io.SeekStart)
+	if diff := cmp.Diff(int64(22), off); diff != "" {
+		t.Errorf("Seek (-want, +got):\n%s", diff)
+	}
+
+	if diff := cmp.Diff(int64(22), z.offset); diff != "" {
+		t.Errorf("r.offset (-want, +got):\n%s", diff)
+	}
+
+	if diff := cmp.Diff(nil, err, cmpopts.EquateErrors()); diff != "" {
+		t.Errorf("Seek (-want, +got):\n%s", diff)
+	}
+
+	buf := make([]byte, 10)
+	n, err := z.Read(buf)
+	if diff := cmp.Diff(0, n); diff != "" {
+		t.Errorf("Read (-want, +got):\n%s", diff)
+	}
+
+	// If Seek goes past the end of the file we return io.EOF.
+	if diff := cmp.Diff(io.EOF, err, cmpopts.EquateErrors()); diff != "" {
+		t.Errorf("Read: %v", err)
+	}
+
+	// z.offset should not be advanced.
+	if diff := cmp.Diff(int64(22), z.offset); diff != "" {
+		t.Errorf("r.offset (-want, +got):\n%s", diff)
+	}
+}
+
+func TestReader_Seek_EOF_past_last_chunk(t *testing.T) {
+	t.Parallel()
+
+	z, err := NewReader(bytes.NewReader([]byte{
+		// Header
+		hdrGzipID1,
+		hdrGzipID2,
+		hdrDeflateCM,
+		flgEXTRA,               // FLG
+		0x00, 0x00, 0x00, 0x00, // MTIME
+		0x2,       // XFL
+		OSUnknown, // OS
+
+		// EXTRA
+		0xa, 0x0, // XLEN // 10
+		0x52, 0x41, // 'R', 'A'
+		0x6, 0x0, // LEN // 6
+		0x1, 0x0, // VER // 1
+		0x5, 0x0, // CHLEN // 5
+		0x0, 0x0, // CHCNT // 0
+
+		0x3, 0x0, 0x0, // Empty deflate data.
+
+		0x0, 0x0, 0x0, 0x0, // CRC32
+		0x0, 0x0, 0x0, 0x0, // ISIZE
+	}))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// NOTE: Seek past the last chunk.
+	off, err := z.Seek(10, io.SeekStart)
+	if diff := cmp.Diff(int64(10), off); diff != "" {
+		t.Errorf("Seek (-want, +got):\n%s", diff)
+	}
+
+	if diff := cmp.Diff(int64(10), z.offset); diff != "" {
+		t.Errorf("r.offset (-want, +got):\n%s", diff)
+	}
+
+	if diff := cmp.Diff(nil, err, cmpopts.EquateErrors()); diff != "" {
+		t.Errorf("Seek (-want, +got):\n%s", diff)
+	}
+
+	buf := make([]byte, 10)
+	n, err := z.Read(buf)
+	if diff := cmp.Diff(0, n); diff != "" {
+		t.Errorf("Read (-want, +got):\n%s", diff)
+	}
+
+	// If Seek goes past the end of the file we return io.EOF.
+	if diff := cmp.Diff(io.EOF, err, cmpopts.EquateErrors()); diff != "" {
+		t.Fatalf("Read: %v", err)
+	}
+
+	// z.offset should not be advanced.
+	if diff := cmp.Diff(int64(10), z.offset); diff != "" {
+		t.Errorf("r.offset (-want, +got):\n%s", diff)
+	}
+}
+
 func TestReader_Seek_SeekStart_negative(t *testing.T) {
 	t.Parallel()
 

--- a/writer_test.go
+++ b/writer_test.go
@@ -358,7 +358,7 @@ func TestWriter(t *testing.T) {
 				0x52, 0x41, // 'R', 'A'
 				0xe, 0x0, // LEN // 14
 				0x1, 0x0, // VER // 1
-				0x6, 0x0, // CHLEN // 65535
+				0x6, 0x0, // CHLEN // 6
 				0x4, 0x0, // CHCNT // 4
 
 				// Chunk sizes.
@@ -404,7 +404,7 @@ func TestWriter(t *testing.T) {
 				0x52, 0x41, // 'R', 'A'
 				0xe, 0x0, // LEN // 14
 				0x1, 0x0, // VER // 1
-				0x6, 0x0, // CHLEN // 65535
+				0x6, 0x0, // CHLEN // 6
 				0x4, 0x0, // CHCNT // 4
 
 				// Chunk sizes.

